### ISSUE TITLE
Add ISO8601 parsing support to qx.land.normalize.Date (BZ#7131)


### ISSUE THREAD

**Issue by qx-bug-importer** [qooxdoo/qooxdoo#7175]
_George Nikolaidis (gnikolaidis) wrote:_

In our projects we are using ISO-8601 dates extensively. We have created a small bootstrap class to enhance the Date class with ISO features.

Since JavaScript 1.8.5 added support in Date.parse() for ISO date parsing, 

https://developer.mozilla.org/en-US/docs/JavaScript/Reference/Global_Objects/Date/parse

we propose including code from (for example)

https://github.com/csnover/js-iso8601/blob/master/iso8601.js

into qx.lang.normalize.Date

We kindly ask Qooxdoo developers to weigh the cost of adding more code to the core framework against the benefits of having said functionality available at all times, and decide if the enhancement proposal is worth considering.

_assigned to Richard Sternagel (@rsternagel)_

**Comment by gnikolaidis** [comment]
Please consider reopening

**Comment by cajus** [comment]
Would you mind creating a pull request for this feature? Sounds like you already have some code available.

**Comment by gnikolaidis** [comment]
Sure, coming up soon

**Comment by cajus** [comment]
Cool. Thanks! Doh. Forgot to re-open.

**Comment by schnepe2** [comment]
Hello again,

I've had time to look over this again and I would like share my thoughts:
What about a user configuration switch (for `config.json`) that lets the application developer decide what behavior is preferred. This would allow the original PR poster to keep his application code unchanged.
Something like `qx.date.parse.mode`, which can be set to either `"standard"` or `"chrome"`  defaulting to `"standard"`
The normalizing code ("`qx.lang.normalize.Date`") can than check what needs to be done to fulfill the "users wish" like:
```
qx.Bootstrap.define("qx.lang.normalize.Date", {

  statics : {

    parse_standard : function (dateString) {
      // ...
    },

    parse_chrome : function (dateString) {
      // ...
    }
  },


  defer : function (statics)
  {
    // ...

    //
    // Date.parse
    //

    // Users "wish"
    var polyfill = qx.core.Environment.get("qx.date.parse.mode");
    // Native "situation"
    var hasParse = qx.core.Environment.get("ecmascript.date.parse");
    var isStd = hasParse && qx.core.Environment.get("ecmascript.date.parse.isstandard");

    switch ( hasParse+"-"+isStd )
    {
      case "false-false": break; // take "standard" or "chrome" polyfill
      case "false-true" : break;// should never happen ;)
      case "true-false" : if (polyfill == "chrome") { polyfill = null; } break;
      case "true-true"  : if (polyfill != "chrome") { polyfill = null; } break;
    }
    if (polyfill) // polyfill on need
    {
      Date.originalParse = Date.parse || function (dateString) { return NaN; };
      Date.parse = statics["parse_"+polyfill];
    }
  }
});
```

The Environment checks are:
```
x.Bootstrap.define("qx.bom.client.EcmaScript",
{
  statics :
  {
    /**
     * Checks if 'parse' supports ISO-8601 parsing
     * @internal
     * @return {Boolean} <code>true</code>, if the method supports ISO-8601 dates.
     */
    getDateParse : function () {
      return Date.parse("1970-01-01T00:00:00Z") === 0;
    },


    /**
     * Checks if 'parse' takes ISO-8601 date strings without timezone specifier
     * and treats them as local (as per specification)
     * @internal
     * @return {Boolean} <code>true</code>, if the method supports ISO-8601 dates.
     */
    getDateParseIsStandard : function () {
      return Date.parse("2001-02-03T04:05:06.007") != Date.parse("2001-02-03T04:05:06.007Z");
       // && Date.parse("2001-02-03T00:00:00Z") == Date.parse("2001-02-03"); // treat 'date only' as UTC
    },

  },


  defer : function(statics) {
    // ...

    // date polyfill
    qx.core.Environment.add("ecmascript.date.now", statics.getDateNow);
    qx.core.Environment.add("ecmascript.date.parse", statics.getDateParse);
    qx.core.Environment.add("ecmascript.date.parse.isstandard", statics.getDateParseIsStandard);

    // ...
  }
});
```

What do you guys think of it.
I would prefer to only have the **standard** version, but the code from the PR acts like chrome and seems to be in production use, so why not have both possibilities.
The name for the 2nd -non standard- option "chrome" however might be subject to change :grin:

**Comment by schnepe2** [comment]
Shouldn't this issue be closed as it is replaced by all the PRs #9112, #9134, #9145?

**Comment by level420** [comment]
@schnepe2 you're right. thank you for reminding.
Fixed by PRs #9112, #9134, #9145
